server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/event/ComputeMessagePreviewListener.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.james.mailbox.events.Group;
+import org.apache.james.mailbox.events.MailboxListener;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.FetchGroupImpl;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.model.MessageResult;
+import org.apache.james.mailbox.store.SessionProvider;
+import org.apache.james.mime4j.dom.Message;
+import org.apache.james.util.mime.MessageContentExtractor.MessageContent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+
+import reactor.core.publisher.Flux;
+import reactor.util.function.Tuple2;
+import reactor.util.function.Tuples;
+
+public class ComputeMessagePreviewListener implements MailboxListener.GroupMailboxListener {
+    public static class ComputeMessagePreviewListenerGroup extends Group {
+
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ComputeMessagePreviewListener.class);
+    private static final Group GROUP = new ComputeMessagePreviewListenerGroup();
+
+    private final MessageIdManager messageIdManager;
+    private final MessagePreviewStore messagePreviewStore;
+    private final SessionProvider sessionProvider;
+    private final MessageExtractorHelper messageExtractorHelper;
+
+    @Inject
+    public ComputeMessagePreviewListener(SessionProvider sessionProvider, MessageIdManager messageIdManager,
+                                            MessagePreviewStore messagePreviewStore, MessageExtractorHelper messageExtractorHelper) {
+        this.sessionProvider = sessionProvider;
+        this.messageIdManager = messageIdManager;
+        this.messagePreviewStore = messagePreviewStore;
+        this.messageExtractorHelper = messageExtractorHelper;
+    }
+
+    @Override
+    public Group getDefaultGroup() {
+        return GROUP;
+    }
+
+    @Override
+    public void event(Event event) throws MailboxException {
+        if (event instanceof Added) {
+            MailboxSession session = sessionProvider.createSystemSession(event.getUsername());
+            handleAddedEvent((Added) event, session);
+        }
+    }
+
+    private void handleAddedEvent(Added addedEvent, MailboxSession session) throws MailboxException {
+        List<MessageId> messageIds = addedEvent.getUids()
+            .stream()
+            .map(uid -> addedEvent.getMetaData(uid).getMessageId())
+            .distinct()
+            .collect(Guavate.toImmutableList());
+
+        preComputePreview(messageIds, session);
+    }
+
+    private void preComputePreview(List<MessageId> messageIds, MailboxSession session) throws MailboxException {
+        Flux.fromIterable(messageIdManager.getMessages(messageIds, FetchGroupImpl.BODY_CONTENT, session))
+            .map(Throwing.function((MessageResult messageResult) -> computePreview(messageResult)))
+            .onErrorContinue((throwable, message) ->
```

Comment:
I'm not OK with that error handling: errors should go in dead-letter (so be thrown here) to enable cheap recomputation in the close future. This is critical.

We should be best effort as much as we can IE ideally we continue processing in the face of error in order to store subsequent previews  - while still reporting the original failure... Note that this is "nice to have" and I can leave without it.

Reply:
In my understanding, or we do throw at failure, then the flux will stop but the all event will end up in dead letters where an admin can redeliver it later, or like here we log the error, continue the treatment... but you might be right that this does not end up in dead letters at all.

I don't see how to get around that suggestion though, I'm not sure if it's possible or if I just cannot see the solution...

Reply:
Please remove this harmful `onErrorContinue` if you don't know how to do better then, you are preventing **deadLetter** to do it's job

Reply:
and as said in the first commit of this PR, you don't have any test regarding this behavior